functorch/functorch/_src/remat_utils_mincut.py

```diff
@@ -0,0 +1,447 @@
+        loc += 1
+
+    def get_capacity(node):
+        # if rematerialize an internal node, need to read and write
+        # might not need to add the write cost, because it might be read by other
+        # might not need to add the read cost, if already reading it - no need the cost
+        user_names_set = set({n.name for n in node_users_map[node.name]})
+        user_names_outside_set = user_names_set.difference(orig_node_names)
+        write_cost = 0  # cost for both read and write because only dest_module is using it
+        if weight and user_names_outside_set.issubset(set(dest_node_names)):
+            write_cost = weight
+
+        read_cost = weight
+
+        capacity = write_cost + read_cost
+        return capacity
+
+    for node in module_origin.graph.nodes:
+        if node.op == 'output':
+            continue
+
+        weight = get_weight(node)
+
+        if ban_recomputation(node):
+            nx_graph.add_edge("source", node.name + "_out", capacity=math.inf)
+
+        # some ops like cuda_batch_norm return tuples, and they cannot be returned as output
+        # because torch.jit.script does not accept
+        # need to return getitem, these getitems might already in the graph
+        # neeed to change the capacity between _in and _out of these ndoes to inf
+        for user in node.users:
+            if user.target == operator.getitem:
+                weight = math.inf
+            break
+
+        if node.op == 'placeholder':
+            capacity = weight
+            nx_graph.add_edge("source", node.name + "_in", capacity=math.inf)
+        elif node.op == 'call_function':
+            capacity = get_capacity(node)
+
+        if (node.name in dest_placeholder_names or
+           (node.name in getitem_users and getitem_users[node.name] in dest_placeholder_names)):
+            nx_graph.add_edge(node.name + "_out", 'sink', capacity=capacity)
+
+        nx_graph.add_edge(node.name + "_in", node.name + "_out", capacity=capacity)
+        for user in node.users:
+            if user.op != "output":
+                nx_graph.add_edge(node.name + "_out", user.name + "_in", capacity=math.inf)
+
+    cut_value, partition = nx.minimum_cut(nx_graph, "source", "sink")
+
+    cut_at_sink = 0
+    for e in nx_graph.edges.data():
+        if e[1] == "sink":
+            cut_at_sink += e[2]["capacity"]
+
+    local_memory_reduced = cut_at_sink - cut_value
+    cut_nodes = get_cut_nodes_from_partition(partition, nx_graph)
+
+    return partition, cut_nodes, local_memory_reduced
+
+
+def check_remat(partition):
+    _, non_reachable = partition
+    return non_reachable != {"sink"}
+
+
+def get_fused_graph(traced_graph):
+    supported_ops = NvFuserOperatorSupport()
+    partitioner = CapabilityBasedPartitioner(traced_graph, supported_ops)
+    fused_graph = partitioner.partition_and_fuse()
+    return fused_graph
+
+
+def rematerialize_fused_graph(fused_graph, node_users_map):
```

Comment:
We should have a description of the algorithm here.

Reply:
added